src/lib.rs

```diff
@@ -29,6 +29,89 @@ mod bits;
 mod memcmp;
 
 use memchr::memchr;
+use std::rc::Rc;
+use std::sync::Arc;
+
+/// Needle that can be searched for within a haystack. It allows specialized
+/// searcher implementations for needle sizes known at compile time.
+pub trait Needle {
+    /// Set to `Some(<usize>)` if and only if the needle's length is known at compile time.
+    const SIZE: Option<usize>;
+    /// Return the slice corresponding to the needle.
+    fn as_bytes(&self) -> &[u8];
+}
+
+impl<const N: usize> Needle for [u8; N] {
+    const SIZE: Option<usize> = Some(N);
+
+    #[inline]
+    fn as_bytes(&self) -> &[u8] {
+        self
+    }
+}
+
+impl Needle for [u8] {
+    const SIZE: Option<usize> = None;
+
+    #[inline]
+    fn as_bytes(&self) -> &[u8] {
+        self
+    }
+}
+
+impl<N: Needle + ?Sized> Needle for Box<N> {
+    const SIZE: Option<usize> = N::SIZE;
+
+    #[inline]
+    fn as_bytes(&self) -> &[u8] {
+        (**self).as_bytes()
+    }
+}
+
+impl<N: Needle + ?Sized> Needle for Rc<N> {
+    const SIZE: Option<usize> = N::SIZE;
+
+    #[inline]
+    fn as_bytes(&self) -> &[u8] {
+        (**self).as_bytes()
+    }
+}
+
+impl<N: Needle + ?Sized> Needle for Arc<N> {
+    const SIZE: Option<usize> = N::SIZE;
+
+    #[inline]
+    fn as_bytes(&self) -> &[u8] {
+        (**self).as_bytes()
+    }
+}
+
+impl<const N: usize> Needle for &[u8; N] {
+    const SIZE: Option<usize> = Some(N);
+
+    #[inline]
+    fn as_bytes(&self) -> &[u8] {
+        *self
+    }
+}
+
+impl Needle for Vec<u8> {
+    const SIZE: Option<usize> = None;
+
+    #[inline]
+    fn as_bytes(&self) -> &[u8] {
+        self
+    }
+}
+
+impl Needle for &[u8] {
+    const SIZE: Option<usize> = None;
+
+    #[inline]
+    fn as_bytes(&self) -> &[u8] {
+        self
+    }
+}
```

Comment:
It doesn't really make a difference, but you can do the same thing for references I think.

```rust
impl<N: Needle + ?Sized> Needle for &N {
    const SIZE: Option<usize> = N::SIZE;

    #[inline]
    fn as_bytes(&self) -> &[u8] {
        (**self).as_bytes()
    }
}
```

Reply:
Good suggestion!